### ISSUE TITLE
wasmer-js Handling Input Output example does not work

### ISSUE THREAD

**Issue by ChrisWhealy** [wasmerio/docs.wasmer.io#20]
First, before I claim there's a bug, are there any undocumented prerequisites needed to get this example working?  For instance, should `duktape` work without first needing to `wapm install` it?

When running the coding shown [here](https://docs.wasmer.io/wasmer-js/node-modules/examples/handling-input-output/wasmer-js-node-modules-handling-input-output), I get a pop-up prompt as expected and if I enter a numeric value, the prompt reappears.  If I then cancel the prompt, the expected value is displayed in both the DOM and the console.

Odd behaviour, but OK...

If I enter any non-numeric value however, the following stack trace appears in the console without any values displayed in either the DOM or the console

```
index.js:119 RuntimeError: unreachable
    at abort (wasm-function[1250]:0xe7264)
    at duk_err_longjmp (wasm-function[440]:0x3629e)
    at duk_err_create_and_throw (wasm-function[23]:0x159a)
    at duk_err_handle_error_fmt (wasm-function[205]:0x1623d)
    at duk__getvar_helper (wasm-function[700]:0x63b40)
    at duk_js_getvar_activation (wasm-function[1173]:0xdbdc2)
    at duk__js_execute_bytecode_inner (wasm-function[1160]:0xd762b)
    at duk_js_execute_bytecode (wasm-function[1146]:0x9dbcb)
    at duk__handle_call_raw (wasm-function[38]:0x2a5d)
    at duk_handle_call_unprotected (wasm-function[33]:0x1d2e)
```

**Comment by torch2424** [comment]
Hello! 

Thanks for opening the issue @ChrisWhealy 😄 !

> First, before I claim there's a bug, are there any undocumented prerequisites needed to get this example working? For instance, should duktape work without first needing to wapm install it?

Not that I know of? Did you follow the steps outlined in [the installlation](https://docs.wasmer.io/wasmer-js/node-modules/wasmer-js-node-modules-installation), [hello world](https://docs.wasmer.io/wasmer-js/node-modules/examples/hello-world/wasmer-js-node-modules-hello-world). And, for this specific example, [Transforming WASI modules](https://docs.wasmer.io/wasmer-js/node-modules/examples/transforming-wasi-modules/wasmer-js-node-modules-transforming-wasi-modules).

> When running the coding shown here, I get a pop-up prompt as expected and if I enter a numeric value, the prompt reappears. If I then cancel the prompt, the expected value is displayed in both the DOM and the console. Odd behaviour, but OK...

Ah yes! The behavior is odd, as it is just for demonstration purposes. Since WASI is a synchronous API, the only way to block the main thread for input, is to use [alert](https://developer.mozilla.org/en-US/docs/Web/API/Window/alert). If you are interested in doing this in a more advanced way, I can write something up, or you can take a look at the [wasm-terminal package](https://github.com/wasmerio/wasmer-js/tree/master/packages/wasm-terminal) in Wasmer JS. 👍 

> If I enter any non-numeric value however, the following stack trace appears in the console without any values displayed in either the DOM or the console

Can you give an example of the value you entered? 😄 If it's valid javascript (like for example, `([0,1,2]).map(x => x * 2)`) should work 😄 

Thank you!

**Comment by ChrisWhealy** [comment]
Hi Aaron

Ok, I've made a new clone of the `docs.wasmer.io` repo and repeated all the installation steps for each of the exercises:

* `hello_world` works fine without modification
* `transforming-wasi-modules` crashes due to a simple typo in line 7 of `index.js`
    ```javascript
    const wasmFilePath = './qjs.wasm';
    ```
    Should read

    ```javascript
    const wasmFilePath = './clock_time_get.wasm';
    ```
    However, would it not be more realistic if `clock_time_get.wasm` actually returned the clock time rather than the hard-coded string `"Done!"`.  I appreciate that the `i32` returned by `$wasi_unstable.clock_time_get` would need to be converted at least to a printable text hex string, but this would at least prove that the function is doing something useful?

* `handling-input-output` works if I enter a numeric value.  But as described above, it only works if I enter a numeric value, press OK, and then press Cancel.  Only then do I see the result.

    Entering ***any*** non-numeric value such as "abc" or "[1,2,3].map(x => x * 2)" always results in the following stack trace:

    ```bash
    RuntimeError: unreachable
        at abort (wasm-function[1250]:0xe7264)
        at duk_err_longjmp (wasm-function[440]:0x3629e)
        at duk_err_create_and_throw (wasm-function[23]:0x159a)
        at duk_err_handle_error (wasm-function[19]:0xf3f)
        at duk__parse_stmt (wasm-function[559]:0x4b095)
        at duk__parse_stmts (wasm-function[507]:0x3fdb0)
        at duk__parse_func_body (wasm-function[498]:0x3dff0)
        at duk__js_compile_raw (wasm-function[486]:0x3cdaf)
        at duk__handle_safe_call_inner (wasm-function[56]:0x4ac2)
        at duk_handle_safe_call (wasm-function[50]:0x456c)
    ```
    As I understand it, `unreachable` generally means some required code is missing.  This could either be due to something like not handling the error side of a Rust `Result` or attempting to call a missing external dependency.  I suspect the latter in this case, but debugging WASM in the browser is nigh on impossible...

:-(

**Comment by torch2424** [comment]
> `hello_world` works fine without modification

Glad that works! 😄 

> `transforming-wasi-modules` crashes due to a simple typo in line 7 of `index.js`

Uh oh! We cna fix that after we get through #23 ? 😄 

> However, would it not be more realistic if clock_time_get.wasm actually returned the clock time rather than the hard-coded string "Done!". I appreciate that the i32 returned by $wasi_unstable.clock_time_get would need to be converted at least to a printable text hex string, but this would at least prove that the function is doing something useful?

Yeah you totally make a good point 😄 Opened: https://github.com/wasmerio/docs.wasmer.io/issues/24 . The reason it was that way, as we made the module as a super simple example for testing the wasm_transformer while we were building it 😄 

> Entering any non-numeric value such as "abc" or "[1,2,3].map(x => x * 2)" always results in the following stack trace:

Ahhhh yeah that should be working 🤔 Maybe I put the wrong duktape module in the example. But thanks for letting us know! I super appreciate it!

> RuntimeError: unreachable

Unreachable can mean anything really. Any type of crash can cause an unreachable unfortunately which makes it more difficult 😮 

> I suspect the latter in this case, but debugging WASM in the browser is nigh on impossible...

So one thing that might help if you keep running into debugging issues with wasm. Try compiling your WebAssembly in like a Debug mode, or using Source maps. I do agree, trying to debug compiled production WebAssembly is a nightmare, but coming from AssemblyScript, or Rust, usually can do some type of logging, or compile into some type of debug mode to greatly improve your experience.

Thanks! 😄 👍
